<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8"/>
    <title>report.html</title>
    <link href="assets/style.css" rel="stylesheet" type="text/css"/></head>
  <body onLoad="init()">
    <script>/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */


function toArray(iter) {
    if (iter === null) {
        return null;
    }
    return Array.prototype.slice.call(iter);
}

function find(selector, elem) { // eslint-disable-line no-redeclare
    if (!elem) {
        elem = document;
    }
    return elem.querySelector(selector);
}

function findAll(selector, elem) {
    if (!elem) {
        elem = document;
    }
    return toArray(elem.querySelectorAll(selector));
}

function sortColumn(elem) {
    toggleSortStates(elem);
    const colIndex = toArray(elem.parentNode.childNodes).indexOf(elem);
    let key;
    if (elem.classList.contains('result')) {
        key = keyResult;
    } else if (elem.classList.contains('links')) {
        key = keyLink;
    } else {
        key = keyAlpha;
    }
    sortTable(elem, key(colIndex));
}

function showAllExtras() { // eslint-disable-line no-unused-vars
    findAll('.col-result').forEach(showExtras);
}

function hideAllExtras() { // eslint-disable-line no-unused-vars
    findAll('.col-result').forEach(hideExtras);
}

function showExtras(colresultElem) {
    const extras = colresultElem.parentNode.nextElementSibling;
    const expandcollapse = colresultElem.firstElementChild;
    extras.classList.remove('collapsed');
    expandcollapse.classList.remove('expander');
    expandcollapse.classList.add('collapser');
}

function hideExtras(colresultElem) {
    const extras = colresultElem.parentNode.nextElementSibling;
    const expandcollapse = colresultElem.firstElementChild;
    extras.classList.add('collapsed');
    expandcollapse.classList.remove('collapser');
    expandcollapse.classList.add('expander');
}

function showFilters() {
    let visibleString = getQueryParameter('visible') || 'all';
    visibleString = visibleString.toLowerCase();
    const checkedItems = visibleString.split(',');

    const filterItems = document.getElementsByClassName('filter');
    for (let i = 0; i < filterItems.length; i++) {
        filterItems[i].hidden = false;

        if (visibleString != 'all') {
            filterItems[i].checked = checkedItems.includes(filterItems[i].getAttribute('data-test-result'));
            filterTable(filterItems[i]);
        }
    }
}

function addCollapse() {
    // Add links for show/hide all
    const resulttable = find('table#results-table');
    const showhideall = document.createElement('p');
    showhideall.innerHTML = '<a href="javascript:showAllExtras()">Show all details</a> / ' +
                            '<a href="javascript:hideAllExtras()">Hide all details</a>';
    resulttable.parentElement.insertBefore(showhideall, resulttable);

    // Add show/hide link to each result
    findAll('.col-result').forEach(function(elem) {
        const collapsed = getQueryParameter('collapsed') || 'Passed';
        const extras = elem.parentNode.nextElementSibling;
        const expandcollapse = document.createElement('span');
        if (extras.classList.contains('collapsed')) {
            expandcollapse.classList.add('expander');
        } else if (collapsed.includes(elem.innerHTML)) {
            extras.classList.add('collapsed');
            expandcollapse.classList.add('expander');
        } else {
            expandcollapse.classList.add('collapser');
        }
        elem.appendChild(expandcollapse);

        elem.addEventListener('click', function(event) {
            if (event.currentTarget.parentNode.nextElementSibling.classList.contains('collapsed')) {
                showExtras(event.currentTarget);
            } else {
                hideExtras(event.currentTarget);
            }
        });
    });
}

function getQueryParameter(name) {
    const match = RegExp('[?&]' + name + '=([^&]*)').exec(window.location.search);
    return match && decodeURIComponent(match[1].replace(/\+/g, ' '));
}

function init () { // eslint-disable-line no-unused-vars
    resetSortHeaders();

    addCollapse();

    showFilters();

    sortColumn(find('.initial-sort'));

    findAll('.sortable').forEach(function(elem) {
        elem.addEventListener('click',
            function() {
                sortColumn(elem);
            }, false);
    });
}

function sortTable(clicked, keyFunc) {
    const rows = findAll('.results-table-row');
    const reversed = !clicked.classList.contains('asc');
    const sortedRows = sort(rows, keyFunc, reversed);
    /* Whole table is removed here because browsers acts much slower
     * when appending existing elements.
     */
    const thead = document.getElementById('results-table-head');
    document.getElementById('results-table').remove();
    const parent = document.createElement('table');
    parent.id = 'results-table';
    parent.appendChild(thead);
    sortedRows.forEach(function(elem) {
        parent.appendChild(elem);
    });
    document.getElementsByTagName('BODY')[0].appendChild(parent);
}

function sort(items, keyFunc, reversed) {
    const sortArray = items.map(function(item, i) {
        return [keyFunc(item), i];
    });

    sortArray.sort(function(a, b) {
        const keyA = a[0];
        const keyB = b[0];

        if (keyA == keyB) return 0;

        if (reversed) {
            return keyA < keyB ? 1 : -1;
        } else {
            return keyA > keyB ? 1 : -1;
        }
    });

    return sortArray.map(function(item) {
        const index = item[1];
        return items[index];
    });
}

function keyAlpha(colIndex) {
    return function(elem) {
        return elem.childNodes[1].childNodes[colIndex].firstChild.data.toLowerCase();
    };
}

function keyLink(colIndex) {
    return function(elem) {
        const dataCell = elem.childNodes[1].childNodes[colIndex].firstChild;
        return dataCell == null ? '' : dataCell.innerText.toLowerCase();
    };
}

function keyResult(colIndex) {
    return function(elem) {
        const strings = ['Error', 'Failed', 'Rerun', 'XFailed', 'XPassed',
            'Skipped', 'Passed'];
        return strings.indexOf(elem.childNodes[1].childNodes[colIndex].firstChild.data);
    };
}

function resetSortHeaders() {
    findAll('.sort-icon').forEach(function(elem) {
        elem.parentNode.removeChild(elem);
    });
    findAll('.sortable').forEach(function(elem) {
        const icon = document.createElement('div');
        icon.className = 'sort-icon';
        icon.textContent = 'vvv';
        elem.insertBefore(icon, elem.firstChild);
        elem.classList.remove('desc', 'active');
        elem.classList.add('asc', 'inactive');
    });
}

function toggleSortStates(elem) {
    //if active, toggle between asc and desc
    if (elem.classList.contains('active')) {
        elem.classList.toggle('asc');
        elem.classList.toggle('desc');
    }

    //if inactive, reset all other functions and add ascending active
    if (elem.classList.contains('inactive')) {
        resetSortHeaders();
        elem.classList.remove('inactive');
        elem.classList.add('active');
    }
}

function isAllRowsHidden(value) {
    return value.hidden == false;
}

function filterTable(elem) { // eslint-disable-line no-unused-vars
    const outcomeAtt = 'data-test-result';
    const outcome = elem.getAttribute(outcomeAtt);
    const classOutcome = outcome + ' results-table-row';
    const outcomeRows = document.getElementsByClassName(classOutcome);

    for(let i = 0; i < outcomeRows.length; i++){
        outcomeRows[i].hidden = !elem.checked;
    }

    const rows = findAll('.results-table-row').filter(isAllRowsHidden);
    const allRowsHidden = rows.length == 0 ? true : false;
    const notFoundMessage = document.getElementById('not-found-message');
    notFoundMessage.hidden = !allRowsHidden;
}
</script>
    <h1>report.html</h1>
    <p>Report generated on 29-Mar-2023 at 08:48:42 by <a href="https://pypi.python.org/pypi/pytest-html">pytest-html</a> v3.2.0</p>
    <h2>Environment</h2>
    <table id="environment">
      <tr>
        <td>Packages</td>
        <td>{"pluggy": "1.0.0", "pytest": "7.2.2"}</td></tr>
      <tr>
        <td>Platform</td>
        <td>Windows-10-10.0.22621-SP0</td></tr>
      <tr>
        <td>Plugins</td>
        <td>{"html": "3.2.0", "metadata": "2.0.4"}</td></tr>
      <tr>
        <td>Python</td>
        <td>3.8.10</td></tr></table>
    <h2>Summary</h2>
    <p>17 tests ran in 19.18 seconds. </p>
    <p class="filter" hidden="true">(Un)check the boxes to filter the results.</p><input checked="true" class="filter" data-test-result="passed" hidden="true" name="filter_checkbox" onChange="filterTable(this)" type="checkbox"/><span class="passed">6 passed</span>, <input checked="true" class="filter" data-test-result="skipped" disabled="true" hidden="true" name="filter_checkbox" onChange="filterTable(this)" type="checkbox"/><span class="skipped">0 skipped</span>, <input checked="true" class="filter" data-test-result="failed" hidden="true" name="filter_checkbox" onChange="filterTable(this)" type="checkbox"/><span class="failed">11 failed</span>, <input checked="true" class="filter" data-test-result="error" hidden="true" name="filter_checkbox" onChange="filterTable(this)" type="checkbox"/><span class="error">2 errors</span>, <input checked="true" class="filter" data-test-result="xfailed" disabled="true" hidden="true" name="filter_checkbox" onChange="filterTable(this)" type="checkbox"/><span class="xfailed">0 expected failures</span>, <input checked="true" class="filter" data-test-result="xpassed" disabled="true" hidden="true" name="filter_checkbox" onChange="filterTable(this)" type="checkbox"/><span class="xpassed">0 unexpected passes</span>
    <h2>Results</h2>
    <table id="results-table">
      <thead id="results-table-head">
        <tr>
          <th class="sortable result initial-sort" col="result">Result</th>
          <th class="sortable" col="name">Test</th>
          <th class="sortable" col="duration">Duration</th>
          <th class="sortable links" col="links">Links</th></tr>
        <tr hidden="true" id="not-found-message">
          <th colspan="4">No results found. Try to check the filters</th></tr></thead>
      <tbody class="error results-table-row">
        <tr>
          <td class="col-result">Error</td>
          <td class="col-name">test_models.py::test_announcement_creation::setup</td>
          <td class="col-duration">0.04</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">self = &lt;sqlalchemy.engine.base.Connection object at 0x000001D8A17D6C40&gt;<br/>dialect = &lt;sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite object at 0x000001D8A1775670&gt;<br/>constructor = &lt;bound method DefaultExecutionContext._init_compiled of &lt;class &#x27;sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext&#x27;&gt;&gt;<br/>statement = &#x27;INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)&#x27;<br/>parameters = (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, ...)<br/>execution_options = immutabledict({&#x27;autocommit&#x27;: True, &#x27;compiled_cache&#x27;: {(&lt;sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite obj...role&#x27;, &#x27;username&#x27;), False, False), &lt;sqlalchemy.dialects.sqlite.base.SQLiteCompiler object at 0x000001D8A17D6EB0&gt;, 1]}})<br/>args = (&lt;sqlalchemy.dialects.sqlite.base.SQLiteCompiler object at 0x000001D8A17D6EB0&gt;, [{&#x27;email&#x27;: &#x27;test_lecturer@abuad.com&#x27;, ...&#x27;last_logged_in_at&#x27;: None, &#x27;password&#x27;: &#x27;password&#x27;, ...}], &lt;sqlalchemy.sql.dml.Insert object at 0x000001D8A17D6B20&gt;, [])<br/>kw = {&#x27;cache_hit&#x27;: symbol(&#x27;CACHE_MISS&#x27;)}, branched = &lt;sqlalchemy.engine.base.Connection object at 0x000001D8A17D6C40&gt;, yp = None<br/>conn = &lt;sqlalchemy.pool.base._ConnectionFairy object at 0x000001D8A17D6E20&gt;<br/>context = &lt;sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext object at 0x000001D8A182B160&gt;<br/>cursor = &lt;sqlite3.Cursor object at 0x000001D8A1818260&gt;<br/>fn = &lt;function ConnectionEvents._listen.&lt;locals&gt;.wrap_before_cursor_execute at 0x000001D8A18045E0&gt;, evt_handled = False<br/><br/>    def _execute_context(<br/>        self,<br/>        dialect,<br/>        constructor,<br/>        statement,<br/>        parameters,<br/>        execution_options,<br/>        *args,<br/>        **kw<br/>    ):<br/>        &quot;&quot;&quot;Create an :class:`.ExecutionContext` and execute, returning<br/>        a :class:`_engine.CursorResult`.&quot;&quot;&quot;<br/>    <br/>        branched = self<br/>        if self.__branch_from:<br/>            # if this is a &quot;branched&quot; connection, do everything in terms<br/>            # of the &quot;root&quot; connection, *except* for .close(), which is<br/>            # the only feature that branching provides<br/>            self = self.__branch_from<br/>    <br/>        if execution_options:<br/>            yp = execution_options.get(&quot;yield_per&quot;, None)<br/>            if yp:<br/>                execution_options = execution_options.union(<br/>                    {&quot;stream_results&quot;: True, &quot;max_row_buffer&quot;: yp}<br/>                )<br/>    <br/>        try:<br/>            conn = self._dbapi_connection<br/>            if conn is None:<br/>                conn = self._revalidate_connection()<br/>    <br/>            context = constructor(<br/>                dialect, self, conn, execution_options, *args, **kw<br/>            )<br/>        except (exc.PendingRollbackError, exc.ResourceClosedError):<br/>            raise<br/>        except BaseException as e:<br/>            self._handle_dbapi_exception(<br/>                e, util.text_type(statement), parameters, None, None<br/>            )<br/>    <br/>        if (<br/>            self._transaction<br/>            and not self._transaction.is_active<br/>            or (<br/>                self._nested_transaction<br/>                and not self._nested_transaction.is_active<br/>            )<br/>        ):<br/>            self._invalid_transaction()<br/>    <br/>        elif self._trans_context_manager:<br/>            TransactionalContext._trans_ctx_check(self)<br/>    <br/>        if self._is_future and self._transaction is None:<br/>            self._autobegin()<br/>    <br/>        context.pre_exec()<br/>    <br/>        if dialect.use_setinputsizes:<br/>            context._set_input_sizes()<br/>    <br/>        cursor, statement, parameters = (<br/>            context.cursor,<br/>            context.statement,<br/>            context.parameters,<br/>        )<br/>    <br/>        if not context.executemany:<br/>            parameters = parameters[0]<br/>    <br/>        if self._has_events or self.engine._has_events:<br/>            for fn in self.dispatch.before_cursor_execute:<br/>                statement, parameters = fn(<br/>                    self,<br/>                    cursor,<br/>                    statement,<br/>                    parameters,<br/>                    context,<br/>                    context.executemany,<br/>                )<br/>    <br/>        if self._echo:<br/>    <br/>            self._log_info(statement)<br/>    <br/>            stats = context._get_cache_stats()<br/>    <br/>            if not self.engine.hide_parameters:<br/>                self._log_info(<br/>                    &quot;[%s] %r&quot;,<br/>                    stats,<br/>                    sql_util._repr_params(<br/>                        parameters, batches=10, ismulti=context.executemany<br/>                    ),<br/>                )<br/>            else:<br/>                self._log_info(<br/>                    &quot;[%s] [SQL parameters hidden due to hide_parameters=True]&quot;<br/>                    % (stats,)<br/>                )<br/>    <br/>        evt_handled = False<br/>        try:<br/>            if context.executemany:<br/>                if self.dialect._has_events:<br/>                    for fn in self.dialect.dispatch.do_executemany:<br/>                        if fn(cursor, statement, parameters, context):<br/>                            evt_handled = True<br/>                            break<br/>                if not evt_handled:<br/>                    self.dialect.do_executemany(<br/>                        cursor, statement, parameters, context<br/>                    )<br/>            elif not parameters and context.no_parameters:<br/>                if self.dialect._has_events:<br/>                    for fn in self.dialect.dispatch.do_execute_no_params:<br/>                        if fn(cursor, statement, context):<br/>                            evt_handled = True<br/>                            break<br/>                if not evt_handled:<br/>                    self.dialect.do_execute_no_params(<br/>                        cursor, statement, context<br/>                    )<br/>            else:<br/>                if self.dialect._has_events:<br/>                    for fn in self.dialect.dispatch.do_execute:<br/>                        if fn(cursor, statement, parameters, context):<br/>                            evt_handled = True<br/>                            break<br/>                if not evt_handled:<br/>&gt;                   self.dialect.do_execute(<br/>                        cursor, statement, parameters, context<br/>                    )<br/><br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:1900: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite object at 0x000001D8A1775670&gt;<br/>cursor = &lt;sqlite3.Cursor object at 0x000001D8A1818260&gt;<br/>statement = &#x27;INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)&#x27;<br/>parameters = (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, ...)<br/>context = &lt;sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext object at 0x000001D8A182B160&gt;<br/><br/>    def do_execute(self, cursor, statement, parameters, context=None):<br/>&gt;       cursor.execute(statement, parameters)<br/><span class="error">E       sqlite3.IntegrityError: UNIQUE constraint failed: lecturers.email</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\engine\default.py:736: IntegrityError<br/><br/>The above exception was the direct cause of the following exception:<br/><br/>    @pytest.fixture(scope=&quot;module&quot;)<br/>    def new_announcement():<br/>        lecturer = Lecturers(<br/>            username=&quot;test_lecturer&quot;,<br/>            email=&quot;test_lecturer@abuad.com&quot;,<br/>            password=&quot;password&quot;,<br/>            role=&quot;lecturer&quot;,<br/>        )<br/>        db.session.add(lecturer)<br/>&gt;       db.session.commit()<br/><br/>test_models.py:30: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/>&lt;string&gt;:2: in commit<br/>    ???<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1451: in commit<br/>    self._transaction.commit(_to_root=self.future)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:829: in commit<br/>    self._prepare_impl()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:808: in _prepare_impl<br/>    self.session.flush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3444: in flush<br/>    self._flush(objects)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3584: in _flush<br/>    transaction.rollback(_capture_exception=True)<br/>..\..\env\lib\site-packages\sqlalchemy\util\langhelpers.py:70: in __exit__<br/>    compat.raise_(<br/>..\..\env\lib\site-packages\sqlalchemy\util\compat.py:211: in raise_<br/>    raise exception<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3544: in _flush<br/>    flush_context.execute()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\unitofwork.py:456: in execute<br/>    rec.execute(self)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\unitofwork.py:630: in execute<br/>    util.preloaded.orm_persistence.save_obj(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\persistence.py:245: in save_obj<br/>    _emit_insert_statements(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\persistence.py:1238: in _emit_insert_statements<br/>    result = connection._execute_20(<br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:1705: in _execute_20<br/>    return meth(self, args_10style, kwargs_10style, execution_options)<br/>..\..\env\lib\site-packages\sqlalchemy\sql\elements.py:334: in _execute_on_connection<br/>    return connection._execute_clauseelement(<br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:1572: in _execute_clauseelement<br/>    ret = self._execute_context(<br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:1943: in _execute_context<br/>    self._handle_dbapi_exception(<br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:2124: in _handle_dbapi_exception<br/>    util.raise_(<br/>..\..\env\lib\site-packages\sqlalchemy\util\compat.py:211: in raise_<br/>    raise exception<br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:1900: in _execute_context<br/>    self.dialect.do_execute(<br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite object at 0x000001D8A1775670&gt;<br/>cursor = &lt;sqlite3.Cursor object at 0x000001D8A1818260&gt;<br/>statement = &#x27;INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)&#x27;<br/>parameters = (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, ...)<br/>context = &lt;sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext object at 0x000001D8A182B160&gt;<br/><br/>    def do_execute(self, cursor, statement, parameters, context=None):<br/>&gt;       cursor.execute(statement, parameters)<br/><span class="error">E       sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: lecturers.email</span><br/><span class="error">E       [SQL: INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)]</span><br/><span class="error">E       [parameters: (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, None, 1, None)]</span><br/><span class="error">E       (Background on this error at: https://sqlalche.me/e/14/gkpj)</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\engine\default.py:736: IntegrityError<br/></div></td></tr></tbody>
      <tbody class="error results-table-row">
        <tr>
          <td class="col-result">Error</td>
          <td class="col-name">test_models.py::test_time_diff_property::setup</td>
          <td class="col-duration">0.00</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">self = &lt;sqlalchemy.engine.base.Connection object at 0x000001D8A17D6C40&gt;<br/>dialect = &lt;sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite object at 0x000001D8A1775670&gt;<br/>constructor = &lt;bound method DefaultExecutionContext._init_compiled of &lt;class &#x27;sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext&#x27;&gt;&gt;<br/>statement = &#x27;INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)&#x27;<br/>parameters = (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, ...)<br/>execution_options = immutabledict({&#x27;autocommit&#x27;: True, &#x27;compiled_cache&#x27;: {(&lt;sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite obj...role&#x27;, &#x27;username&#x27;), False, False), &lt;sqlalchemy.dialects.sqlite.base.SQLiteCompiler object at 0x000001D8A17D6EB0&gt;, 1]}})<br/>args = (&lt;sqlalchemy.dialects.sqlite.base.SQLiteCompiler object at 0x000001D8A17D6EB0&gt;, [{&#x27;email&#x27;: &#x27;test_lecturer@abuad.com&#x27;, ...&#x27;last_logged_in_at&#x27;: None, &#x27;password&#x27;: &#x27;password&#x27;, ...}], &lt;sqlalchemy.sql.dml.Insert object at 0x000001D8A17D6B20&gt;, [])<br/>kw = {&#x27;cache_hit&#x27;: symbol(&#x27;CACHE_MISS&#x27;)}, branched = &lt;sqlalchemy.engine.base.Connection object at 0x000001D8A17D6C40&gt;, yp = None<br/>conn = &lt;sqlalchemy.pool.base._ConnectionFairy object at 0x000001D8A17D6E20&gt;<br/>context = &lt;sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext object at 0x000001D8A182B160&gt;<br/>cursor = &lt;sqlite3.Cursor object at 0x000001D8A1818260&gt;<br/>fn = &lt;function ConnectionEvents._listen.&lt;locals&gt;.wrap_before_cursor_execute at 0x000001D8A18045E0&gt;, evt_handled = False<br/><br/>    def _execute_context(<br/>        self,<br/>        dialect,<br/>        constructor,<br/>        statement,<br/>        parameters,<br/>        execution_options,<br/>        *args,<br/>        **kw<br/>    ):<br/>        &quot;&quot;&quot;Create an :class:`.ExecutionContext` and execute, returning<br/>        a :class:`_engine.CursorResult`.&quot;&quot;&quot;<br/>    <br/>        branched = self<br/>        if self.__branch_from:<br/>            # if this is a &quot;branched&quot; connection, do everything in terms<br/>            # of the &quot;root&quot; connection, *except* for .close(), which is<br/>            # the only feature that branching provides<br/>            self = self.__branch_from<br/>    <br/>        if execution_options:<br/>            yp = execution_options.get(&quot;yield_per&quot;, None)<br/>            if yp:<br/>                execution_options = execution_options.union(<br/>                    {&quot;stream_results&quot;: True, &quot;max_row_buffer&quot;: yp}<br/>                )<br/>    <br/>        try:<br/>            conn = self._dbapi_connection<br/>            if conn is None:<br/>                conn = self._revalidate_connection()<br/>    <br/>            context = constructor(<br/>                dialect, self, conn, execution_options, *args, **kw<br/>            )<br/>        except (exc.PendingRollbackError, exc.ResourceClosedError):<br/>            raise<br/>        except BaseException as e:<br/>            self._handle_dbapi_exception(<br/>                e, util.text_type(statement), parameters, None, None<br/>            )<br/>    <br/>        if (<br/>            self._transaction<br/>            and not self._transaction.is_active<br/>            or (<br/>                self._nested_transaction<br/>                and not self._nested_transaction.is_active<br/>            )<br/>        ):<br/>            self._invalid_transaction()<br/>    <br/>        elif self._trans_context_manager:<br/>            TransactionalContext._trans_ctx_check(self)<br/>    <br/>        if self._is_future and self._transaction is None:<br/>            self._autobegin()<br/>    <br/>        context.pre_exec()<br/>    <br/>        if dialect.use_setinputsizes:<br/>            context._set_input_sizes()<br/>    <br/>        cursor, statement, parameters = (<br/>            context.cursor,<br/>            context.statement,<br/>            context.parameters,<br/>        )<br/>    <br/>        if not context.executemany:<br/>            parameters = parameters[0]<br/>    <br/>        if self._has_events or self.engine._has_events:<br/>            for fn in self.dispatch.before_cursor_execute:<br/>                statement, parameters = fn(<br/>                    self,<br/>                    cursor,<br/>                    statement,<br/>                    parameters,<br/>                    context,<br/>                    context.executemany,<br/>                )<br/>    <br/>        if self._echo:<br/>    <br/>            self._log_info(statement)<br/>    <br/>            stats = context._get_cache_stats()<br/>    <br/>            if not self.engine.hide_parameters:<br/>                self._log_info(<br/>                    &quot;[%s] %r&quot;,<br/>                    stats,<br/>                    sql_util._repr_params(<br/>                        parameters, batches=10, ismulti=context.executemany<br/>                    ),<br/>                )<br/>            else:<br/>                self._log_info(<br/>                    &quot;[%s] [SQL parameters hidden due to hide_parameters=True]&quot;<br/>                    % (stats,)<br/>                )<br/>    <br/>        evt_handled = False<br/>        try:<br/>            if context.executemany:<br/>                if self.dialect._has_events:<br/>                    for fn in self.dialect.dispatch.do_executemany:<br/>                        if fn(cursor, statement, parameters, context):<br/>                            evt_handled = True<br/>                            break<br/>                if not evt_handled:<br/>                    self.dialect.do_executemany(<br/>                        cursor, statement, parameters, context<br/>                    )<br/>            elif not parameters and context.no_parameters:<br/>                if self.dialect._has_events:<br/>                    for fn in self.dialect.dispatch.do_execute_no_params:<br/>                        if fn(cursor, statement, context):<br/>                            evt_handled = True<br/>                            break<br/>                if not evt_handled:<br/>                    self.dialect.do_execute_no_params(<br/>                        cursor, statement, context<br/>                    )<br/>            else:<br/>                if self.dialect._has_events:<br/>                    for fn in self.dialect.dispatch.do_execute:<br/>                        if fn(cursor, statement, parameters, context):<br/>                            evt_handled = True<br/>                            break<br/>                if not evt_handled:<br/>&gt;                   self.dialect.do_execute(<br/>                        cursor, statement, parameters, context<br/>                    )<br/><br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:1900: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite object at 0x000001D8A1775670&gt;<br/>cursor = &lt;sqlite3.Cursor object at 0x000001D8A1818260&gt;<br/>statement = &#x27;INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)&#x27;<br/>parameters = (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, ...)<br/>context = &lt;sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext object at 0x000001D8A182B160&gt;<br/><br/>    def do_execute(self, cursor, statement, parameters, context=None):<br/>&gt;       cursor.execute(statement, parameters)<br/><span class="error">E       sqlite3.IntegrityError: UNIQUE constraint failed: lecturers.email</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\engine\default.py:736: IntegrityError<br/><br/>The above exception was the direct cause of the following exception:<br/><br/>    @pytest.fixture(scope=&quot;module&quot;)<br/>    def new_announcement():<br/>        lecturer = Lecturers(<br/>            username=&quot;test_lecturer&quot;,<br/>            email=&quot;test_lecturer@abuad.com&quot;,<br/>            password=&quot;password&quot;,<br/>            role=&quot;lecturer&quot;,<br/>        )<br/>        db.session.add(lecturer)<br/>&gt;       db.session.commit()<br/><br/>test_models.py:30: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/>&lt;string&gt;:2: in commit<br/>    ???<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1451: in commit<br/>    self._transaction.commit(_to_root=self.future)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:829: in commit<br/>    self._prepare_impl()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:808: in _prepare_impl<br/>    self.session.flush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3444: in flush<br/>    self._flush(objects)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3584: in _flush<br/>    transaction.rollback(_capture_exception=True)<br/>..\..\env\lib\site-packages\sqlalchemy\util\langhelpers.py:70: in __exit__<br/>    compat.raise_(<br/>..\..\env\lib\site-packages\sqlalchemy\util\compat.py:211: in raise_<br/>    raise exception<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3544: in _flush<br/>    flush_context.execute()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\unitofwork.py:456: in execute<br/>    rec.execute(self)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\unitofwork.py:630: in execute<br/>    util.preloaded.orm_persistence.save_obj(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\persistence.py:245: in save_obj<br/>    _emit_insert_statements(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\persistence.py:1238: in _emit_insert_statements<br/>    result = connection._execute_20(<br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:1705: in _execute_20<br/>    return meth(self, args_10style, kwargs_10style, execution_options)<br/>..\..\env\lib\site-packages\sqlalchemy\sql\elements.py:334: in _execute_on_connection<br/>    return connection._execute_clauseelement(<br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:1572: in _execute_clauseelement<br/>    ret = self._execute_context(<br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:1943: in _execute_context<br/>    self._handle_dbapi_exception(<br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:2124: in _handle_dbapi_exception<br/>    util.raise_(<br/>..\..\env\lib\site-packages\sqlalchemy\util\compat.py:211: in raise_<br/>    raise exception<br/>..\..\env\lib\site-packages\sqlalchemy\engine\base.py:1900: in _execute_context<br/>    self.dialect.do_execute(<br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite object at 0x000001D8A1775670&gt;<br/>cursor = &lt;sqlite3.Cursor object at 0x000001D8A1818260&gt;<br/>statement = &#x27;INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)&#x27;<br/>parameters = (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, ...)<br/>context = &lt;sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext object at 0x000001D8A182B160&gt;<br/><br/>    def do_execute(self, cursor, statement, parameters, context=None):<br/>&gt;       cursor.execute(statement, parameters)<br/><span class="error">E       sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: lecturers.email</span><br/><span class="error">E       [SQL: INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)]</span><br/><span class="error">E       [parameters: (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, None, 1, None)]</span><br/><span class="error">E       (Background on this error at: https://sqlalche.me/e/14/gkpj)</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\engine\default.py:736: IntegrityError<br/></div></td></tr></tbody>
      <tbody class="failed results-table-row">
        <tr>
          <td class="col-result">Failed</td>
          <td class="col-name">test_models.py::test_student_creation</td>
          <td class="col-duration">0.00</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">def test_student_creation():<br/>        password = generate_password_hash(&quot;password&quot;, method=&quot;sha256&quot;)<br/>        student = Students(<br/>            username=&quot;john_doe&quot;,<br/>            email=&quot;john_doe@example.com&quot;,<br/>            password=password,<br/>            matric_number=&quot;MAT123456&quot;,<br/>            department=&quot;Computer Science&quot;,<br/>            role=&quot;student&quot;,<br/>        )<br/>        db.session.add(student)<br/>&gt;       db.session.commit()<br/><br/>test_models.py:68: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/>&lt;string&gt;:2: in commit<br/>    ???<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1451: in commit<br/>    self._transaction.commit(_to_root=self.future)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:827: in commit<br/>    self._assert_active(prepared_ok=True)<br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.orm.session.SessionTransaction object at 0x000001D8A1775A00&gt;, prepared_ok = True, rollback_ok = False, deactive_ok = False<br/>closed_msg = &#x27;This transaction is closed&#x27;<br/><br/>    def _assert_active(<br/>        self,<br/>        prepared_ok=False,<br/>        rollback_ok=False,<br/>        deactive_ok=False,<br/>        closed_msg=&quot;This transaction is closed&quot;,<br/>    ):<br/>        if self._state is COMMITTED:<br/>            raise sa_exc.InvalidRequestError(<br/>                &quot;This session is in &#x27;committed&#x27; state; no further &quot;<br/>                &quot;SQL can be emitted within this transaction.&quot;<br/>            )<br/>        elif self._state is PREPARED:<br/>            if not prepared_ok:<br/>                raise sa_exc.InvalidRequestError(<br/>                    &quot;This session is in &#x27;prepared&#x27; state; no further &quot;<br/>                    &quot;SQL can be emitted within this transaction.&quot;<br/>                )<br/>        elif self._state is DEACTIVE:<br/>            if not deactive_ok and not rollback_ok:<br/>                if self._rollback_exception:<br/>&gt;                   raise sa_exc.PendingRollbackError(<br/>                        &quot;This Session&#x27;s transaction has been rolled back &quot;<br/>                        &quot;due to a previous exception during flush.&quot;<br/>                        &quot; To begin a new transaction with this Session, &quot;<br/>                        &quot;first issue Session.rollback().&quot;<br/>                        &quot; Original exception was: %s&quot;<br/>                        % self._rollback_exception,<br/>                        code=&quot;7s2a&quot;,<br/><span class="error">E                       sqlalchemy.exc.PendingRollbackError: This Session&#x27;s transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (sqlite3.IntegrityError) UNIQUE constraint failed: lecturers.email</span><br/><span class="error">E                       [SQL: INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)]</span><br/><span class="error">E                       [parameters: (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, None, 1, None)]</span><br/><span class="error">E                       (Background on this error at: https://sqlalche.me/e/14/gkpj) (Background on this error at: https://sqlalche.me/e/14/7s2a)</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:601: PendingRollbackError<br/></div></td></tr></tbody>
      <tbody class="failed results-table-row">
        <tr>
          <td class="col-result">Failed</td>
          <td class="col-name">test_models.py::test_lecturer_creation</td>
          <td class="col-duration">0.00</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">def test_lecturer_creation():<br/>        password = generate_password_hash(&quot;password&quot;, method=&quot;sha256&quot;)<br/>    <br/>        lecturer = Lecturers(<br/>            username=&quot;jane_doe&quot;,<br/>            email=&quot;jane_doe@example.com&quot;,<br/>            password=password,<br/>            role=&quot;lecturer&quot;,<br/>        )<br/>        db.session.add(lecturer)<br/>&gt;       db.session.commit()<br/><br/>test_models.py:82: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/>&lt;string&gt;:2: in commit<br/>    ???<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1451: in commit<br/>    self._transaction.commit(_to_root=self.future)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:827: in commit<br/>    self._assert_active(prepared_ok=True)<br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.orm.session.SessionTransaction object at 0x000001D8A1775A00&gt;, prepared_ok = True, rollback_ok = False, deactive_ok = False<br/>closed_msg = &#x27;This transaction is closed&#x27;<br/><br/>    def _assert_active(<br/>        self,<br/>        prepared_ok=False,<br/>        rollback_ok=False,<br/>        deactive_ok=False,<br/>        closed_msg=&quot;This transaction is closed&quot;,<br/>    ):<br/>        if self._state is COMMITTED:<br/>            raise sa_exc.InvalidRequestError(<br/>                &quot;This session is in &#x27;committed&#x27; state; no further &quot;<br/>                &quot;SQL can be emitted within this transaction.&quot;<br/>            )<br/>        elif self._state is PREPARED:<br/>            if not prepared_ok:<br/>                raise sa_exc.InvalidRequestError(<br/>                    &quot;This session is in &#x27;prepared&#x27; state; no further &quot;<br/>                    &quot;SQL can be emitted within this transaction.&quot;<br/>                )<br/>        elif self._state is DEACTIVE:<br/>            if not deactive_ok and not rollback_ok:<br/>                if self._rollback_exception:<br/>&gt;                   raise sa_exc.PendingRollbackError(<br/>                        &quot;This Session&#x27;s transaction has been rolled back &quot;<br/>                        &quot;due to a previous exception during flush.&quot;<br/>                        &quot; To begin a new transaction with this Session, &quot;<br/>                        &quot;first issue Session.rollback().&quot;<br/>                        &quot; Original exception was: %s&quot;<br/>                        % self._rollback_exception,<br/>                        code=&quot;7s2a&quot;,<br/><span class="error">E                       sqlalchemy.exc.PendingRollbackError: This Session&#x27;s transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (sqlite3.IntegrityError) UNIQUE constraint failed: lecturers.email</span><br/><span class="error">E                       [SQL: INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)]</span><br/><span class="error">E                       [parameters: (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, None, 1, None)]</span><br/><span class="error">E                       (Background on this error at: https://sqlalche.me/e/14/gkpj) (Background on this error at: https://sqlalche.me/e/14/7s2a)</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:601: PendingRollbackError<br/></div></td></tr></tbody>
      <tbody class="failed results-table-row">
        <tr>
          <td class="col-result">Failed</td>
          <td class="col-name">test_models.py::test_check_password</td>
          <td class="col-duration">0.00</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">def test_check_password():<br/>        password = generate_password_hash(&quot;password&quot;, method=&quot;sha256&quot;)<br/>    <br/>        student = Students(<br/>            username=&quot;john_doe&quot;,<br/>            email=&quot;john_doe@example.com&quot;,<br/>            password=password,<br/>            matric_number=&quot;19/ENG06/098&quot;,<br/>            department=&quot;Computer Engineering&quot;,<br/>            role=&quot;student&quot;,<br/>        )<br/>        db.session.add(student)<br/>&gt;       db.session.commit()<br/><br/>test_models.py:98: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/>&lt;string&gt;:2: in commit<br/>    ???<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1451: in commit<br/>    self._transaction.commit(_to_root=self.future)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:827: in commit<br/>    self._assert_active(prepared_ok=True)<br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.orm.session.SessionTransaction object at 0x000001D8A1775A00&gt;, prepared_ok = True, rollback_ok = False, deactive_ok = False<br/>closed_msg = &#x27;This transaction is closed&#x27;<br/><br/>    def _assert_active(<br/>        self,<br/>        prepared_ok=False,<br/>        rollback_ok=False,<br/>        deactive_ok=False,<br/>        closed_msg=&quot;This transaction is closed&quot;,<br/>    ):<br/>        if self._state is COMMITTED:<br/>            raise sa_exc.InvalidRequestError(<br/>                &quot;This session is in &#x27;committed&#x27; state; no further &quot;<br/>                &quot;SQL can be emitted within this transaction.&quot;<br/>            )<br/>        elif self._state is PREPARED:<br/>            if not prepared_ok:<br/>                raise sa_exc.InvalidRequestError(<br/>                    &quot;This session is in &#x27;prepared&#x27; state; no further &quot;<br/>                    &quot;SQL can be emitted within this transaction.&quot;<br/>                )<br/>        elif self._state is DEACTIVE:<br/>            if not deactive_ok and not rollback_ok:<br/>                if self._rollback_exception:<br/>&gt;                   raise sa_exc.PendingRollbackError(<br/>                        &quot;This Session&#x27;s transaction has been rolled back &quot;<br/>                        &quot;due to a previous exception during flush.&quot;<br/>                        &quot; To begin a new transaction with this Session, &quot;<br/>                        &quot;first issue Session.rollback().&quot;<br/>                        &quot; Original exception was: %s&quot;<br/>                        % self._rollback_exception,<br/>                        code=&quot;7s2a&quot;,<br/><span class="error">E                       sqlalchemy.exc.PendingRollbackError: This Session&#x27;s transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (sqlite3.IntegrityError) UNIQUE constraint failed: lecturers.email</span><br/><span class="error">E                       [SQL: INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)]</span><br/><span class="error">E                       [parameters: (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, None, 1, None)]</span><br/><span class="error">E                       (Background on this error at: https://sqlalche.me/e/14/gkpj) (Background on this error at: https://sqlalche.me/e/14/7s2a)</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:601: PendingRollbackError<br/></div></td></tr></tbody>
      <tbody class="failed results-table-row">
        <tr>
          <td class="col-result">Failed</td>
          <td class="col-name">test_routes.py::test_login_student</td>
          <td class="col-duration">0.04</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">def test_login_student():<br/>        name = Flask(__name__)<br/>    <br/>        name = create_app(name)<br/>    <br/>        client = name.test_client()<br/>    <br/>        # Send a POST request to the login route with form data<br/>&gt;       response_student = client.post(<br/>            &quot;/login&quot;,<br/>            data={&quot;username&quot;: &quot;19/ENG02/077&quot;, &quot;password&quot;: &quot;password&quot;, &quot;role&quot;: &quot;student&quot;},<br/>        )<br/><br/>test_routes.py:20: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/>..\..\env\lib\site-packages\werkzeug\test.py:1136: in post<br/>    return self.open(*args, **kw)<br/>..\..\env\lib\site-packages\flask\testing.py:235: in open<br/>    return super().open(<br/>..\..\env\lib\site-packages\werkzeug\test.py:1076: in open<br/>    response = self.run_wsgi_app(request.environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:945: in run_wsgi_app<br/>    rv = run_wsgi_app(self.application, environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:1233: in run_wsgi_app<br/>    app_rv = app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2091: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask_socketio\__init__.py:45: in __call__<br/>    return super(_SocketIOMiddleware, self).__call__(environ,<br/>..\..\env\lib\site-packages\engineio\middleware.py:74: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2076: in wsgi_app<br/>    response = self.handle_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:2073: in wsgi_app<br/>    response = self.full_dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1518: in full_dispatch_request<br/>    rv = self.handle_user_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:1516: in full_dispatch_request<br/>    rv = self.dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1502: in dispatch_request<br/>    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)<br/>..\views\login.py:30: in show<br/>    user = Students.query.filter_by(matric_number=username).first()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\query.py:2824: in first<br/>    return self.limit(1)._iter().first()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\query.py:2916: in _iter<br/>    result = self.session.execute(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1662: in execute<br/>    ) = compile_state_cls.orm_pre_session_exec(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\context.py:312: in orm_pre_session_exec<br/>    session._autoflush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:2248: in _autoflush<br/>    self.flush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3444: in flush<br/>    self._flush(objects)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3540: in _flush<br/>    flush_context.transaction = transaction = self.begin(_subtrans=True)<br/>&lt;string&gt;:2: in begin<br/>    ???<br/>..\..\env\lib\site-packages\sqlalchemy\util\deprecations.py:309: in warned<br/>    return fn(*args, **kwargs)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1326: in begin<br/>    trans = self._transaction._begin(nested=nested)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:629: in _begin<br/>    self._assert_active()<br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.orm.session.SessionTransaction object at 0x000001D8A1775A00&gt;, prepared_ok = False, rollback_ok = False, deactive_ok = False<br/>closed_msg = &#x27;This transaction is closed&#x27;<br/><br/>    def _assert_active(<br/>        self,<br/>        prepared_ok=False,<br/>        rollback_ok=False,<br/>        deactive_ok=False,<br/>        closed_msg=&quot;This transaction is closed&quot;,<br/>    ):<br/>        if self._state is COMMITTED:<br/>            raise sa_exc.InvalidRequestError(<br/>                &quot;This session is in &#x27;committed&#x27; state; no further &quot;<br/>                &quot;SQL can be emitted within this transaction.&quot;<br/>            )<br/>        elif self._state is PREPARED:<br/>            if not prepared_ok:<br/>                raise sa_exc.InvalidRequestError(<br/>                    &quot;This session is in &#x27;prepared&#x27; state; no further &quot;<br/>                    &quot;SQL can be emitted within this transaction.&quot;<br/>                )<br/>        elif self._state is DEACTIVE:<br/>            if not deactive_ok and not rollback_ok:<br/>                if self._rollback_exception:<br/>&gt;                   raise sa_exc.PendingRollbackError(<br/>                        &quot;This Session&#x27;s transaction has been rolled back &quot;<br/>                        &quot;due to a previous exception during flush.&quot;<br/>                        &quot; To begin a new transaction with this Session, &quot;<br/>                        &quot;first issue Session.rollback().&quot;<br/>                        &quot; Original exception was: %s&quot;<br/>                        % self._rollback_exception,<br/>                        code=&quot;7s2a&quot;,<br/><span class="error">E                       sqlalchemy.exc.PendingRollbackError: This Session&#x27;s transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (sqlite3.IntegrityError) UNIQUE constraint failed: lecturers.email</span><br/><span class="error">E                       [SQL: INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)]</span><br/><span class="error">E                       [parameters: (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, None, 1, None)]</span><br/><span class="error">E                       (Background on this error at: https://sqlalche.me/e/14/gkpj) (Background on this error at: https://sqlalche.me/e/14/7s2a)</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:601: PendingRollbackError<br/></div></td></tr></tbody>
      <tbody class="failed results-table-row">
        <tr>
          <td class="col-result">Failed</td>
          <td class="col-name">test_routes.py::test_student</td>
          <td class="col-duration">0.08</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">def test_student():<br/>        name = Flask(__name__)<br/>    <br/>        name = create_app(name)<br/>    <br/>        client = name.test_client()<br/>    <br/>        # Send a POST request to the login route with form data<br/>&gt;       response_student = client.post(<br/>            &quot;/login&quot;,<br/>            data={&quot;username&quot;: &quot;19/ENG02/077&quot;, &quot;password&quot;: &quot;password&quot;, &quot;role&quot;: &quot;student&quot;},<br/>        )<br/><br/>test_routes.py:38: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/>..\..\env\lib\site-packages\werkzeug\test.py:1136: in post<br/>    return self.open(*args, **kw)<br/>..\..\env\lib\site-packages\flask\testing.py:235: in open<br/>    return super().open(<br/>..\..\env\lib\site-packages\werkzeug\test.py:1076: in open<br/>    response = self.run_wsgi_app(request.environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:945: in run_wsgi_app<br/>    rv = run_wsgi_app(self.application, environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:1233: in run_wsgi_app<br/>    app_rv = app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2091: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask_socketio\__init__.py:45: in __call__<br/>    return super(_SocketIOMiddleware, self).__call__(environ,<br/>..\..\env\lib\site-packages\engineio\middleware.py:74: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2076: in wsgi_app<br/>    response = self.handle_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:2073: in wsgi_app<br/>    response = self.full_dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1518: in full_dispatch_request<br/>    rv = self.handle_user_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:1516: in full_dispatch_request<br/>    rv = self.dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1502: in dispatch_request<br/>    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)<br/>..\views\login.py:30: in show<br/>    user = Students.query.filter_by(matric_number=username).first()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\query.py:2824: in first<br/>    return self.limit(1)._iter().first()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\query.py:2916: in _iter<br/>    result = self.session.execute(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1662: in execute<br/>    ) = compile_state_cls.orm_pre_session_exec(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\context.py:312: in orm_pre_session_exec<br/>    session._autoflush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:2248: in _autoflush<br/>    self.flush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3444: in flush<br/>    self._flush(objects)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3540: in _flush<br/>    flush_context.transaction = transaction = self.begin(_subtrans=True)<br/>&lt;string&gt;:2: in begin<br/>    ???<br/>..\..\env\lib\site-packages\sqlalchemy\util\deprecations.py:309: in warned<br/>    return fn(*args, **kwargs)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1326: in begin<br/>    trans = self._transaction._begin(nested=nested)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:629: in _begin<br/>    self._assert_active()<br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.orm.session.SessionTransaction object at 0x000001D8A1775A00&gt;, prepared_ok = False, rollback_ok = False, deactive_ok = False<br/>closed_msg = &#x27;This transaction is closed&#x27;<br/><br/>    def _assert_active(<br/>        self,<br/>        prepared_ok=False,<br/>        rollback_ok=False,<br/>        deactive_ok=False,<br/>        closed_msg=&quot;This transaction is closed&quot;,<br/>    ):<br/>        if self._state is COMMITTED:<br/>            raise sa_exc.InvalidRequestError(<br/>                &quot;This session is in &#x27;committed&#x27; state; no further &quot;<br/>                &quot;SQL can be emitted within this transaction.&quot;<br/>            )<br/>        elif self._state is PREPARED:<br/>            if not prepared_ok:<br/>                raise sa_exc.InvalidRequestError(<br/>                    &quot;This session is in &#x27;prepared&#x27; state; no further &quot;<br/>                    &quot;SQL can be emitted within this transaction.&quot;<br/>                )<br/>        elif self._state is DEACTIVE:<br/>            if not deactive_ok and not rollback_ok:<br/>                if self._rollback_exception:<br/>&gt;                   raise sa_exc.PendingRollbackError(<br/>                        &quot;This Session&#x27;s transaction has been rolled back &quot;<br/>                        &quot;due to a previous exception during flush.&quot;<br/>                        &quot; To begin a new transaction with this Session, &quot;<br/>                        &quot;first issue Session.rollback().&quot;<br/>                        &quot; Original exception was: %s&quot;<br/>                        % self._rollback_exception,<br/>                        code=&quot;7s2a&quot;,<br/><span class="error">E                       sqlalchemy.exc.PendingRollbackError: This Session&#x27;s transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (sqlite3.IntegrityError) UNIQUE constraint failed: lecturers.email</span><br/><span class="error">E                       [SQL: INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)]</span><br/><span class="error">E                       [parameters: (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, None, 1, None)]</span><br/><span class="error">E                       (Background on this error at: https://sqlalche.me/e/14/gkpj) (Background on this error at: https://sqlalche.me/e/14/7s2a)</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:601: PendingRollbackError<br/></div></td></tr></tbody>
      <tbody class="failed results-table-row">
        <tr>
          <td class="col-result">Failed</td>
          <td class="col-name">test_routes.py::test_login_lecturer</td>
          <td class="col-duration">0.08</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">def test_login_lecturer():<br/>        name = Flask(__name__)<br/>    <br/>        name = create_app(name)<br/>    <br/>        client = name.test_client()<br/>    <br/>        # Send a POST request to the login route with form data<br/>&gt;       response_lecturer = client.post(<br/>            &quot;/login&quot;,<br/>            data={<br/>                &quot;username&quot;: &quot;Adebayo Adigun&quot;,<br/>                &quot;password&quot;: &quot;password&quot;,<br/>                &quot;role&quot;: &quot;lecturer&quot;,<br/>            },<br/>        )<br/><br/>test_routes.py:56: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/>..\..\env\lib\site-packages\werkzeug\test.py:1136: in post<br/>    return self.open(*args, **kw)<br/>..\..\env\lib\site-packages\flask\testing.py:235: in open<br/>    return super().open(<br/>..\..\env\lib\site-packages\werkzeug\test.py:1076: in open<br/>    response = self.run_wsgi_app(request.environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:945: in run_wsgi_app<br/>    rv = run_wsgi_app(self.application, environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:1233: in run_wsgi_app<br/>    app_rv = app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2091: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask_socketio\__init__.py:45: in __call__<br/>    return super(_SocketIOMiddleware, self).__call__(environ,<br/>..\..\env\lib\site-packages\engineio\middleware.py:74: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2076: in wsgi_app<br/>    response = self.handle_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:2073: in wsgi_app<br/>    response = self.full_dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1518: in full_dispatch_request<br/>    rv = self.handle_user_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:1516: in full_dispatch_request<br/>    rv = self.dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1502: in dispatch_request<br/>    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)<br/>..\views\login.py:39: in show<br/>    return render_template(&quot;/pages/login.html&quot;)<br/>..\..\env\lib\site-packages\flask\templating.py:148: in render_template<br/>    ctx.app.jinja_env.get_or_select_template(template_name_or_list),<br/>..\..\env\lib\site-packages\jinja2\environment.py:1081: in get_or_select_template<br/>    return self.get_template(template_name_or_list, parent, globals)<br/>..\..\env\lib\site-packages\jinja2\environment.py:1010: in get_template<br/>    return self._load_template(name, globals)<br/>..\..\env\lib\site-packages\jinja2\environment.py:969: in _load_template<br/>    template = self.loader.load(self, name, self.make_globals(globals))<br/>..\..\env\lib\site-packages\jinja2\loaders.py:126: in load<br/>    source, filename, uptodate = self.get_source(environment, name)<br/>..\..\env\lib\site-packages\flask\templating.py:59: in get_source<br/>    return self._get_source_fast(environment, template)<br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;flask.templating.DispatchingJinjaLoader object at 0x000001D8A2296130&gt;<br/>environment = &lt;flask.templating.Environment object at 0x000001D8A2296040&gt;, template = &#x27;/pages/login.html&#x27;<br/><br/>    def _get_source_fast(<br/>        self, environment: Environment, template: str<br/>    ) -&gt; t.Tuple[str, t.Optional[str], t.Optional[t.Callable]]:<br/>        for _srcobj, loader in self._iter_loaders(template):<br/>            try:<br/>                return loader.get_source(environment, template)<br/>            except TemplateNotFound:<br/>                continue<br/>&gt;       raise TemplateNotFound(template)<br/><span class="error">E       jinja2.exceptions.TemplateNotFound: /pages/login.html</span><br/><br/>..\..\env\lib\site-packages\flask\templating.py:95: TemplateNotFound<br/></div></td></tr></tbody>
      <tbody class="failed results-table-row">
        <tr>
          <td class="col-result">Failed</td>
          <td class="col-name">test_routes.py::test_reset_password_lecturer</td>
          <td class="col-duration">0.12</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">def test_reset_password_lecturer():<br/>        name = Flask(__name__, template_folder=&quot;../templates&quot;)<br/>    <br/>        name = create_app(name)<br/>    <br/>        client = name.test_client()<br/>    <br/>        response_lecturer = client.get(&quot;reset_password&quot;)<br/>        assert response_lecturer.status_code == 200<br/>    <br/>        # Send a POST request to the login route with form data<br/>&gt;       response_lecturer = client.post(<br/>            &quot;/reset_password&quot;,<br/>            data={&quot;mail&quot;: &quot;ifeanyinneji777@gmail.com&quot;},<br/>        )<br/><br/>test_routes.py:81: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/>..\..\env\lib\site-packages\werkzeug\test.py:1136: in post<br/>    return self.open(*args, **kw)<br/>..\..\env\lib\site-packages\flask\testing.py:235: in open<br/>    return super().open(<br/>..\..\env\lib\site-packages\werkzeug\test.py:1076: in open<br/>    response = self.run_wsgi_app(request.environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:945: in run_wsgi_app<br/>    rv = run_wsgi_app(self.application, environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:1233: in run_wsgi_app<br/>    app_rv = app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2091: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask_socketio\__init__.py:45: in __call__<br/>    return super(_SocketIOMiddleware, self).__call__(environ,<br/>..\..\env\lib\site-packages\engineio\middleware.py:74: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2076: in wsgi_app<br/>    response = self.handle_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:2073: in wsgi_app<br/>    response = self.full_dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1518: in full_dispatch_request<br/>    rv = self.handle_user_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:1516: in full_dispatch_request<br/>    rv = self.dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1502: in dispatch_request<br/>    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)<br/>..\views\reset_password.py:27: in reset_password<br/>    student = Students.query.filter_by(email=mail).first()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\query.py:2824: in first<br/>    return self.limit(1)._iter().first()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\query.py:2916: in _iter<br/>    result = self.session.execute(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1662: in execute<br/>    ) = compile_state_cls.orm_pre_session_exec(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\context.py:312: in orm_pre_session_exec<br/>    session._autoflush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:2248: in _autoflush<br/>    self.flush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3444: in flush<br/>    self._flush(objects)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3540: in _flush<br/>    flush_context.transaction = transaction = self.begin(_subtrans=True)<br/>&lt;string&gt;:2: in begin<br/>    ???<br/>..\..\env\lib\site-packages\sqlalchemy\util\deprecations.py:309: in warned<br/>    return fn(*args, **kwargs)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1326: in begin<br/>    trans = self._transaction._begin(nested=nested)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:629: in _begin<br/>    self._assert_active()<br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.orm.session.SessionTransaction object at 0x000001D8A1775A00&gt;, prepared_ok = False, rollback_ok = False, deactive_ok = False<br/>closed_msg = &#x27;This transaction is closed&#x27;<br/><br/>    def _assert_active(<br/>        self,<br/>        prepared_ok=False,<br/>        rollback_ok=False,<br/>        deactive_ok=False,<br/>        closed_msg=&quot;This transaction is closed&quot;,<br/>    ):<br/>        if self._state is COMMITTED:<br/>            raise sa_exc.InvalidRequestError(<br/>                &quot;This session is in &#x27;committed&#x27; state; no further &quot;<br/>                &quot;SQL can be emitted within this transaction.&quot;<br/>            )<br/>        elif self._state is PREPARED:<br/>            if not prepared_ok:<br/>                raise sa_exc.InvalidRequestError(<br/>                    &quot;This session is in &#x27;prepared&#x27; state; no further &quot;<br/>                    &quot;SQL can be emitted within this transaction.&quot;<br/>                )<br/>        elif self._state is DEACTIVE:<br/>            if not deactive_ok and not rollback_ok:<br/>                if self._rollback_exception:<br/>&gt;                   raise sa_exc.PendingRollbackError(<br/>                        &quot;This Session&#x27;s transaction has been rolled back &quot;<br/>                        &quot;due to a previous exception during flush.&quot;<br/>                        &quot; To begin a new transaction with this Session, &quot;<br/>                        &quot;first issue Session.rollback().&quot;<br/>                        &quot; Original exception was: %s&quot;<br/>                        % self._rollback_exception,<br/>                        code=&quot;7s2a&quot;,<br/><span class="error">E                       sqlalchemy.exc.PendingRollbackError: This Session&#x27;s transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (sqlite3.IntegrityError) UNIQUE constraint failed: lecturers.email</span><br/><span class="error">E                       [SQL: INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)]</span><br/><span class="error">E                       [parameters: (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, None, 1, None)]</span><br/><span class="error">E                       (Background on this error at: https://sqlalche.me/e/14/gkpj) (Background on this error at: https://sqlalche.me/e/14/7s2a)</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:601: PendingRollbackError<br/></div></td></tr></tbody>
      <tbody class="failed results-table-row">
        <tr>
          <td class="col-result">Failed</td>
          <td class="col-name">test_routes.py::test_login_admin</td>
          <td class="col-duration">0.07</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">def test_login_admin():<br/>        name = Flask(__name__)<br/>    <br/>        name = create_app(name)<br/>    <br/>        client = name.test_client()<br/>    <br/>        # Send a POST request to the login route with form data<br/>&gt;       response_admin = client.post(<br/>            &quot;/login&quot;,<br/>            data={&quot;username&quot;: &quot;admin&quot;, &quot;password&quot;: &quot;admin&quot;, &quot;role&quot;: &quot;admin&quot;},<br/>        )<br/><br/>test_routes.py:98: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/>..\..\env\lib\site-packages\werkzeug\test.py:1136: in post<br/>    return self.open(*args, **kw)<br/>..\..\env\lib\site-packages\flask\testing.py:235: in open<br/>    return super().open(<br/>..\..\env\lib\site-packages\werkzeug\test.py:1076: in open<br/>    response = self.run_wsgi_app(request.environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:945: in run_wsgi_app<br/>    rv = run_wsgi_app(self.application, environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:1233: in run_wsgi_app<br/>    app_rv = app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2091: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask_socketio\__init__.py:45: in __call__<br/>    return super(_SocketIOMiddleware, self).__call__(environ,<br/>..\..\env\lib\site-packages\engineio\middleware.py:74: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2076: in wsgi_app<br/>    response = self.handle_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:2073: in wsgi_app<br/>    response = self.full_dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1518: in full_dispatch_request<br/>    rv = self.handle_user_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:1516: in full_dispatch_request<br/>    rv = self.dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1502: in dispatch_request<br/>    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)<br/>..\views\login.py:41: in show<br/>    user = Admins.query.filter_by(username=username).first()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\query.py:2824: in first<br/>    return self.limit(1)._iter().first()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\query.py:2916: in _iter<br/>    result = self.session.execute(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1662: in execute<br/>    ) = compile_state_cls.orm_pre_session_exec(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\context.py:312: in orm_pre_session_exec<br/>    session._autoflush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:2248: in _autoflush<br/>    self.flush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3444: in flush<br/>    self._flush(objects)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3540: in _flush<br/>    flush_context.transaction = transaction = self.begin(_subtrans=True)<br/>&lt;string&gt;:2: in begin<br/>    ???<br/>..\..\env\lib\site-packages\sqlalchemy\util\deprecations.py:309: in warned<br/>    return fn(*args, **kwargs)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1326: in begin<br/>    trans = self._transaction._begin(nested=nested)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:629: in _begin<br/>    self._assert_active()<br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.orm.session.SessionTransaction object at 0x000001D8A1775A00&gt;, prepared_ok = False, rollback_ok = False, deactive_ok = False<br/>closed_msg = &#x27;This transaction is closed&#x27;<br/><br/>    def _assert_active(<br/>        self,<br/>        prepared_ok=False,<br/>        rollback_ok=False,<br/>        deactive_ok=False,<br/>        closed_msg=&quot;This transaction is closed&quot;,<br/>    ):<br/>        if self._state is COMMITTED:<br/>            raise sa_exc.InvalidRequestError(<br/>                &quot;This session is in &#x27;committed&#x27; state; no further &quot;<br/>                &quot;SQL can be emitted within this transaction.&quot;<br/>            )<br/>        elif self._state is PREPARED:<br/>            if not prepared_ok:<br/>                raise sa_exc.InvalidRequestError(<br/>                    &quot;This session is in &#x27;prepared&#x27; state; no further &quot;<br/>                    &quot;SQL can be emitted within this transaction.&quot;<br/>                )<br/>        elif self._state is DEACTIVE:<br/>            if not deactive_ok and not rollback_ok:<br/>                if self._rollback_exception:<br/>&gt;                   raise sa_exc.PendingRollbackError(<br/>                        &quot;This Session&#x27;s transaction has been rolled back &quot;<br/>                        &quot;due to a previous exception during flush.&quot;<br/>                        &quot; To begin a new transaction with this Session, &quot;<br/>                        &quot;first issue Session.rollback().&quot;<br/>                        &quot; Original exception was: %s&quot;<br/>                        % self._rollback_exception,<br/>                        code=&quot;7s2a&quot;,<br/><span class="error">E                       sqlalchemy.exc.PendingRollbackError: This Session&#x27;s transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (sqlite3.IntegrityError) UNIQUE constraint failed: lecturers.email</span><br/><span class="error">E                       [SQL: INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)]</span><br/><span class="error">E                       [parameters: (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, None, 1, None)]</span><br/><span class="error">E                       (Background on this error at: https://sqlalche.me/e/14/gkpj) (Background on this error at: https://sqlalche.me/e/14/7s2a)</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:601: PendingRollbackError<br/></div></td></tr></tbody>
      <tbody class="failed results-table-row">
        <tr>
          <td class="col-result">Failed</td>
          <td class="col-name">test_routes.py::test_logout</td>
          <td class="col-duration">0.08</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">def test_logout():<br/>        name = Flask(__name__)<br/>    <br/>        name = create_app(name)<br/>    <br/>        client = name.test_client()<br/>    <br/>        # Send a POST request to the login route with form data<br/>&gt;       response = client.post(<br/>            &quot;/login&quot;,<br/>            data={&quot;username&quot;: &quot;admin&quot;, &quot;password&quot;: &quot;admin&quot;, &quot;role&quot;: &quot;admin&quot;},<br/>        )<br/><br/>test_routes.py:116: <br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/>..\..\env\lib\site-packages\werkzeug\test.py:1136: in post<br/>    return self.open(*args, **kw)<br/>..\..\env\lib\site-packages\flask\testing.py:235: in open<br/>    return super().open(<br/>..\..\env\lib\site-packages\werkzeug\test.py:1076: in open<br/>    response = self.run_wsgi_app(request.environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:945: in run_wsgi_app<br/>    rv = run_wsgi_app(self.application, environ, buffered=buffered)<br/>..\..\env\lib\site-packages\werkzeug\test.py:1233: in run_wsgi_app<br/>    app_rv = app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2091: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask_socketio\__init__.py:45: in __call__<br/>    return super(_SocketIOMiddleware, self).__call__(environ,<br/>..\..\env\lib\site-packages\engineio\middleware.py:74: in __call__<br/>    return self.wsgi_app(environ, start_response)<br/>..\..\env\lib\site-packages\flask\app.py:2076: in wsgi_app<br/>    response = self.handle_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:2073: in wsgi_app<br/>    response = self.full_dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1518: in full_dispatch_request<br/>    rv = self.handle_user_exception(e)<br/>..\..\env\lib\site-packages\flask\app.py:1516: in full_dispatch_request<br/>    rv = self.dispatch_request()<br/>..\..\env\lib\site-packages\flask\app.py:1502: in dispatch_request<br/>    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)<br/>..\views\login.py:41: in show<br/>    user = Admins.query.filter_by(username=username).first()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\query.py:2824: in first<br/>    return self.limit(1)._iter().first()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\query.py:2916: in _iter<br/>    result = self.session.execute(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1662: in execute<br/>    ) = compile_state_cls.orm_pre_session_exec(<br/>..\..\env\lib\site-packages\sqlalchemy\orm\context.py:312: in orm_pre_session_exec<br/>    session._autoflush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:2248: in _autoflush<br/>    self.flush()<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3444: in flush<br/>    self._flush(objects)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:3540: in _flush<br/>    flush_context.transaction = transaction = self.begin(_subtrans=True)<br/>&lt;string&gt;:2: in begin<br/>    ???<br/>..\..\env\lib\site-packages\sqlalchemy\util\deprecations.py:309: in warned<br/>    return fn(*args, **kwargs)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:1326: in begin<br/>    trans = self._transaction._begin(nested=nested)<br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:629: in _begin<br/>    self._assert_active()<br/>_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ <br/><br/>self = &lt;sqlalchemy.orm.session.SessionTransaction object at 0x000001D8A1775A00&gt;, prepared_ok = False, rollback_ok = False, deactive_ok = False<br/>closed_msg = &#x27;This transaction is closed&#x27;<br/><br/>    def _assert_active(<br/>        self,<br/>        prepared_ok=False,<br/>        rollback_ok=False,<br/>        deactive_ok=False,<br/>        closed_msg=&quot;This transaction is closed&quot;,<br/>    ):<br/>        if self._state is COMMITTED:<br/>            raise sa_exc.InvalidRequestError(<br/>                &quot;This session is in &#x27;committed&#x27; state; no further &quot;<br/>                &quot;SQL can be emitted within this transaction.&quot;<br/>            )<br/>        elif self._state is PREPARED:<br/>            if not prepared_ok:<br/>                raise sa_exc.InvalidRequestError(<br/>                    &quot;This session is in &#x27;prepared&#x27; state; no further &quot;<br/>                    &quot;SQL can be emitted within this transaction.&quot;<br/>                )<br/>        elif self._state is DEACTIVE:<br/>            if not deactive_ok and not rollback_ok:<br/>                if self._rollback_exception:<br/>&gt;                   raise sa_exc.PendingRollbackError(<br/>                        &quot;This Session&#x27;s transaction has been rolled back &quot;<br/>                        &quot;due to a previous exception during flush.&quot;<br/>                        &quot; To begin a new transaction with this Session, &quot;<br/>                        &quot;first issue Session.rollback().&quot;<br/>                        &quot; Original exception was: %s&quot;<br/>                        % self._rollback_exception,<br/>                        code=&quot;7s2a&quot;,<br/><span class="error">E                       sqlalchemy.exc.PendingRollbackError: This Session&#x27;s transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (sqlite3.IntegrityError) UNIQUE constraint failed: lecturers.email</span><br/><span class="error">E                       [SQL: INSERT INTO lecturers (id, username, email, password, role, created_at, last_logged_in_at, is_active, &quot;hashCode&quot;) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)]</span><br/><span class="error">E                       [parameters: (&#x27;69da0c4e-9a5b-4fa6-8814-a0cafd26e665&#x27;, &#x27;test_lecturer&#x27;, &#x27;test_lecturer@abuad.com&#x27;, &#x27;password&#x27;, &#x27;lecturer&#x27;, &#x27;2023-03-29 07:48:31.700577&#x27;, None, 1, None)]</span><br/><span class="error">E                       (Background on this error at: https://sqlalche.me/e/14/gkpj) (Background on this error at: https://sqlalche.me/e/14/7s2a)</span><br/><br/>..\..\env\lib\site-packages\sqlalchemy\orm\session.py:601: PendingRollbackError<br/></div></td></tr></tbody>
      <tbody class="failed results-table-row">
        <tr>
          <td class="col-result">Failed</td>
          <td class="col-name">test_routes.py::test_lecturer_record_attendance_route</td>
          <td class="col-duration">0.11</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">def test_lecturer_record_attendance_route():<br/>        name = Flask(__name__, template_folder=&quot;../templates&quot;)<br/>        name = create_app(name)<br/>        client = name.test_client()<br/>        response_lecturer = client.post(<br/>            &quot;/login&quot;,<br/>            data={<br/>                &quot;username&quot;: &quot;Afolabi Gbadamosi&quot;,<br/>                &quot;password&quot;: &quot;password&quot;,<br/>                &quot;role&quot;: &quot;lecturer&quot;,<br/>            },<br/>        )<br/>        response_lecturer = client.get(&quot;/lecturer/record_attendance/501&quot;)<br/>&gt;       assert response_lecturer.status_code == 200<br/><span class="error">E       assert 401 == 200</span><br/><span class="error">E        +  where 401 = &lt;WrapperTestResponse streamed [401 UNAUTHORIZED]&gt;.status_code</span><br/><br/>test_routes.py:155: AssertionError<br/></div></td></tr></tbody>
      <tbody class="failed results-table-row">
        <tr>
          <td class="col-result">Failed</td>
          <td class="col-name">test_routes.py::test_lecturer_view_student_route</td>
          <td class="col-duration">0.08</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log">def test_lecturer_view_student_route():<br/>        name = Flask(__name__, template_folder=&quot;../templates&quot;)<br/>        name = create_app(name)<br/>        client = name.test_client()<br/>        response_lecturer = client.post(<br/>            &quot;/login&quot;,<br/>            data={<br/>                &quot;username&quot;: &quot;Afolabi Gbadamosi&quot;,<br/>                &quot;password&quot;: &quot;password&quot;,<br/>                &quot;role&quot;: &quot;lecturer&quot;,<br/>            },<br/>        )<br/>        response_lecturer = client.get(&quot;/lecturer/view_students/501&quot;)<br/>&gt;       assert response_lecturer.status_code == 200<br/><span class="error">E       assert 401 == 200</span><br/><span class="error">E        +  where 401 = &lt;WrapperTestResponse streamed [401 UNAUTHORIZED]&gt;.status_code</span><br/><br/>test_routes.py:171: AssertionError<br/></div></td></tr></tbody>
      <tbody class="passed results-table-row">
        <tr>
          <td class="col-result">Passed</td>
          <td class="col-name">test_functions.py::test_validate_matric_number</td>
          <td class="col-duration">0.00</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="empty log">No log output captured.</div></td></tr></tbody>
      <tbody class="passed results-table-row">
        <tr>
          <td class="col-result">Passed</td>
          <td class="col-name">test_functions.py::test_send_mail</td>
          <td class="col-duration">3.89</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="log"> ------------------------------Captured stderr call------------------------------ <br/>send: &#x27;ehlo [192.168.50.52]\r\n&#x27;
reply: b&#x27;250-smtp.mailtrap.io\r\n&#x27;
reply: b&#x27;250-SIZE 5242880\r\n&#x27;
reply: b&#x27;250-PIPELINING\r\n&#x27;
reply: b&#x27;250-ENHANCEDSTATUSCODES\r\n&#x27;
reply: b&#x27;250-8BITMIME\r\n&#x27;
reply: b&#x27;250-DSN\r\n&#x27;
reply: b&#x27;250-AUTH PLAIN LOGIN CRAM-MD5\r\n&#x27;
reply: b&#x27;250 STARTTLS\r\n&#x27;
reply: retcode (250); Msg: b&#x27;smtp.mailtrap.io\nSIZE 5242880\nPIPELINING\nENHANCEDSTATUSCODES\n8BITMIME\nDSN\nAUTH PLAIN LOGIN CRAM-MD5\nSTARTTLS&#x27;
send: &#x27;STARTTLS\r\n&#x27;
reply: b&#x27;220 2.0.0 Start TLS\r\n&#x27;
reply: retcode (220); Msg: b&#x27;2.0.0 Start TLS&#x27;
send: &#x27;ehlo [192.168.50.52]\r\n&#x27;
reply: b&#x27;250-smtp.mailtrap.io\r\n&#x27;
reply: b&#x27;250-SIZE 5242880\r\n&#x27;
reply: b&#x27;250-PIPELINING\r\n&#x27;
reply: b&#x27;250-ENHANCEDSTATUSCODES\r\n&#x27;
reply: b&#x27;250-8BITMIME\r\n&#x27;
reply: b&#x27;250-DSN\r\n&#x27;
reply: b&#x27;250 AUTH PLAIN LOGIN CRAM-MD5\r\n&#x27;
reply: retcode (250); Msg: b&#x27;smtp.mailtrap.io\nSIZE 5242880\nPIPELINING\nENHANCEDSTATUSCODES\n8BITMIME\nDSN\nAUTH PLAIN LOGIN CRAM-MD5&#x27;
send: &#x27;AUTH CRAM-MD5\r\n&#x27;
reply: b&#x27;334 PDc5MDg2MDU5Ni4xNjgwMDc2MTEwQHNtdHAubWFpbHRyYXAuaW8+\r\n&#x27;
reply: retcode (334); Msg: b&#x27;PDc5MDg2MDU5Ni4xNjgwMDc2MTEwQHNtdHAubWFpbHRyYXAuaW8+&#x27;
send: &#x27;OTVhYTMyZGQ1Yjk5NTUgYmQyMTJjMTg1ODBkMzJjNGU5NGViNjQzMzI1NWNkMTQ=\r\n&#x27;
reply: b&#x27;235 2.0.0 OK\r\n&#x27;
reply: retcode (235); Msg: b&#x27;2.0.0 OK&#x27;
send: &#x27;mail FROM:&lt;95aa32dd5b9955&gt; size=842\r\n&#x27;
reply: b&#x27;250 2.1.0 Ok\r\n&#x27;
reply: retcode (250); Msg: b&#x27;2.1.0 Ok&#x27;
send: &#x27;rcpt TO:&lt;test@example.com&gt;\r\n&#x27;
reply: b&#x27;250 2.1.0 Ok\r\n&#x27;
reply: retcode (250); Msg: b&#x27;2.1.0 Ok&#x27;
send: &#x27;data\r\n&#x27;
reply: b&#x27;354 Go ahead\r\n&#x27;
reply: retcode (354); Msg: b&#x27;Go ahead&#x27;
data: (354, b&#x27;Go ahead&#x27;)
send: b&#x27;Content-Type: multipart/mixed; boundary=&quot;===============7811522838871212709==&quot;\r\nMIME-Version: 1.0\r\nSubject: Test Subject\r\nFrom: 95aa32dd5b9955\r\nTo: test@example.com\r\nDate: Wed, 29 Mar 2023 08:48:30 +0100\r\nMessage-ID: &lt;168007610775.19044.14759414567265532206@LAPTOP-T1HPDDBK&gt;\r\n\r\n--===============7811522838871212709==\r\nContent-Type: multipart/alternative;\r\n boundary=&quot;===============0744702162959165955==&quot;\r\nMIME-Version: 1.0\r\n\r\n--===============0744702162959165955==\r\nContent-Type: text/plain; charset=&quot;utf-8&quot;\r\nMIME-Version: 1.0\r\nContent-Transfer-Encoding: 7bit\r\n\r\n\r\n--===============0744702162959165955==\r\nContent-Type: text/html; charset=&quot;utf-8&quot;\r\nMIME-Version: 1.0\r\nContent-Transfer-Encoding: 7bit\r\n\r\n&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;Test HTML\r\n\r\n&lt;/body&gt;&lt;/html&gt;\r\n--===============0744702162959165955==--\r\n\r\n--===============7811522838871212709==--\r\n.\r\n&#x27;
reply: b&#x27;250 2.0.0 Ok: queued\r\n&#x27;
reply: retcode (250); Msg: b&#x27;2.0.0 Ok: queued&#x27;
data: (250, b&#x27;2.0.0 Ok: queued&#x27;)
send: &#x27;quit\r\n&#x27;
reply: b&#x27;221 2.0.0 Bye\r\n&#x27;
reply: retcode (221); Msg: b&#x27;2.0.0 Bye&#x27;
<br/></div></td></tr></tbody>
      <tbody class="passed results-table-row">
        <tr>
          <td class="col-result">Passed</td>
          <td class="col-name">test_functions.py::test_base64_to_image</td>
          <td class="col-duration">0.03</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="empty log">No log output captured.</div></td></tr></tbody>
      <tbody class="passed results-table-row">
        <tr>
          <td class="col-result">Passed</td>
          <td class="col-name">test_functions.py::test_count_names_in_files</td>
          <td class="col-duration">0.00</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="empty log">No log output captured.</div></td></tr></tbody>
      <tbody class="passed results-table-row">
        <tr>
          <td class="col-result">Passed</td>
          <td class="col-name">test_routes.py::test_index</td>
          <td class="col-duration">0.10</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="empty log">No log output captured.</div></td></tr></tbody>
      <tbody class="passed results-table-row">
        <tr>
          <td class="col-result">Passed</td>
          <td class="col-name">test_routes.py::test_lecturer_register_student_route</td>
          <td class="col-duration">0.08</td>
          <td class="col-links"></td></tr>
        <tr>
          <td class="extra" colspan="4">
            <div class="empty log">No log output captured.</div></td></tr></tbody></table></body></html>